File: Src/Compilers/VisualBasic/Source/Syntax/Annotation/SyntaxAnnotation.TriviaResolver.vb
#If False Then
Imports System.Collections.Generic
Imports Roslyn.Compilers.Internal

Namespace Roslyn.Compilers.VisualBasic
    Partial Public Class SyntaxAnnotation
        Private Class TriviaResolver
            Inherits SyntaxWalker(Of List(Of SyntaxTrivia))
            Private ReadOnly annotation As SyntaxAnnotation

            Public Shared Function Resolve(root As SyntaxNode, annotation As SyntaxAnnotation) As IEnumerable(Of SyntaxTrivia)
                Contract.ThrowIfNull(root)
                Contract.ThrowIfNull(annotation)

                Dim result = New List(Of SyntaxTrivia)()
                Dim resolver = New TriviaResolver(annotation)
                resolver.Visit(root, result)

                Return result
            End Function

            Private Sub New(annotation As SyntaxAnnotation)
                MyBase.New(visitIntoStructuredTrivia:=True)
                Me.annotation = annotation
            End Sub

            Public Overrides Function Visit(node As SyntaxNode, Optional results As List(Of SyntaxTrivia) = Nothing) As Object
                ' if it doesnt have annotations, don't even bother to go in.
                If Not node.HasAnnotations Then
                    Return Nothing
                End If

                Return MyBase.Visit(node, results)
            End Function

            Public Overrides Sub VisitToken(token As SyntaxToken, results As List(Of SyntaxTrivia))
                ' if it doesnt have annotations, don't even bother to go in.
                If Not token.HasAnnotations Then
                    Return
                End If

                MyBase.VisitToken(token, results)
            End Sub

            Public Overrides Sub VisitTrivia(trivia As SyntaxTrivia, results As List(Of SyntaxTrivia))
                ' if it doesnt have annotations, don't even bother to go in.
                If Not trivia.HasAnnotations Then
                    Return
                End If

                Dim annotations = trivia.UnderlyingNode.GetAnnotations()
                For i As Integer = 0 To annotations.Length - 1
                    If annotations(i) Is annotation Then
                        results.Add(trivia)
                        Exit For
                    End If
                Next i

                MyBase.VisitTrivia(trivia, results)
            End Sub
        End Class
    End Class
End Namespace
#End If

File: Src/Compilers/CSharp/Source/Syntax/Annotation/SyntaxAnnotation.TriviaResolver.cs
#if false
using System.Collections.Generic;
using Roslyn.Compilers.Internal;
using Roslyn.Utilities;

namespace Roslyn.Compilers.CSharp
{
    public partial class SyntaxAnnotation
    {
        private class TriviaResolver : SyntaxWalker<List<SyntaxTrivia>>
        {
            private readonly SyntaxAnnotation annotation;

            public static IEnumerable<SyntaxTrivia> Resolve(SyntaxNode root, SyntaxAnnotation annotation)
            {
                Contract.ThrowIfNull(root);
                Contract.ThrowIfNull(annotation);

                var result = new List<SyntaxTrivia>();
                var resolver = new TriviaResolver(annotation);
                resolver.Visit(root, result);

                return result;
            }

            private TriviaResolver(SyntaxAnnotation annotation) :
                base(visitIntoStructuredTrivia: true)
            {
                this.annotation = annotation;
            }

            public override object Visit(SyntaxNode node, List<SyntaxTrivia> results)
            {
                // if it doesnt have annotations, don't even bother to go in.
                if (!node.HasAnnotations)
                {
                    return null;
                }

                return base.Visit(node, results);
            }

            public override void VisitToken(SyntaxToken token, List<SyntaxTrivia> results)
            {
                // if it doesnt have annotations, don't even bother to go in.
                if (!token.HasAnnotations)
                {
                    return;
                }

                base.VisitToken(token, results);
            }

            public override void VisitTrivia(SyntaxTrivia trivia, List<SyntaxTrivia> results)
            {
                // if it doesnt have annotations, don't even bother to go in.
                if (!trivia.HasAnnotations)
                {
                    return;
                }

                var annotations = trivia.UnderlyingNode.GetAnnotations();
                for (int i = 0; i < annotations.Length; i++)
                {
                    if (annotations[i] == annotation)
                    {
                        results.Add(trivia);
                        break;
                    }
                }

                base.VisitTrivia(trivia, results);
            }
        }
    }
}
#endif